### PR TITLE
feat: allow retention_seconds on append only table and its downstream index

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close https://github.com/risingwavelabs/risingwave/issues/14791

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Users can set TTL(time to live) on the append-only table with the `retention_seconds` option and the record will be removed from the table after the retention seconds passed.
The same retention seconds will be applied to the table's downstream index automatically.
```
dev=> create table t(v int) APPEND ONLY with (retention_seconds = 5);
NOTICE:  APPEND ONLY TABLE is currently an experimental feature.
CREATE_TABLE
dev=> create index i on t(v);
CREATE_INDEX
dev=> insert into t values(1);
INSERT 0 1
dev=> flush;
FLUSH
dev=> select * from t;
 v 
---
 1
(1 row)

dev=> select * from i;
 v 
---
 1
(1 row)

dev=> select pg_sleep(5);
 pg_sleep 
----------
         
(1 row)

dev=> select * from i;
 v 
---
(0 rows)

dev=> select * from t;
 v 
---
(0 rows)
```

</details>

### CODE REVIEW

File: e2e_test/ddl/table/table.slt.part [hunk 1]
```diff
@@ -259,3 +259,42 @@ drop table t1;
...
+----
+1
+
+query I
+select * from i;
+----
+1
+
+statement ok
+select pg_sleep(10);
```
Comment (fuyufjh) [inline e2e_test/ddl/table/table.slt.part:289]: QQ. Once a row has expired, but hasn't been removed by compactor yet, will the scan iterator "hides" it?

**Review by fuyufjh — APPROVED** [review]
LGTM

File: e2e_test/ddl/table/table.slt.part (hunk 1 above)
Comment (st1page) [reply e2e_test/ddl/table/table.slt.part:289]: I remember there is some code in state store implementation to "hide" it. need to check @Li0k

Comment (Li0k) [reply e2e_test/ddl/table/table.slt.part:289]: We definitely hide it, by filtering it in the `min_epoch` of the iterator on the read path.

**Pull request merged**